kernel/src/hil/symmetric_encryption.rs

```diff
@@ -139,3 +139,35 @@ pub trait AES128CCM<'a> {
         encrypting: bool,
     ) -> Result<(), (ErrorCode, &'static mut [u8])>;
 }
+
+pub trait GCMClient {
+    /// `res` is Ok(()) if the encryption/decryption process succeeded. This
+    /// does not mean that the message has been verified in the case of
+    /// decryption.
+    /// If we are encrypting: `tag_is_valid` is `true` iff `res` is Ok(()).
+    /// If we are decrypting: `tag_is_valid` is `true` iff `res` is Ok(()) and the
+    /// message authentication tag is valid.
+    fn crypt_done(&self, buf: &'static mut [u8], res: Result<(), ErrorCode>, tag_is_valid: bool);
+}
+
+pub trait AES128GCM<'a> {
+    /// Set the client instance which will receive `crypt_done()` callbacks
+    fn set_client(&'a self, client: &'a dyn GCMClient);
+
+    /// Set the key to be used for GCM encryption
+    fn set_key(&self, key: &[u8]) -> Result<(), ErrorCode>;
+
+    /// Set the IV to be used for GCM encryption. The IV should be less
+    /// then 12 bytes (96 bits) as recommened in NIST-800-38D.
+    fn set_iv(&self, nonce: &[u8]) -> Result<(), ErrorCode>;
```

Comment:
I believe its less than or equal to 12 bytes, not less than 12 bytes?

Reply:
What are the allowed errors and what triggers them?

Reply:
Shouldn't that be SIZE?

Reply:
It looks like the comment on SIZE is "too large", but I think that was because it didn't consider operations like this. It seems like SIZE should be more general than that. Otherwise, > 12 means SIZE and < 12 means INVAL.

Reply:
Similar to above, we aren't trying to store anything in a buffer I don't think it's a `SIZE` error (the buffer isn't too small), but more that the user passed an incorrect IV

Reply:
I agree -- that makes sense.